Read input data

yjchoi1 opened the pull request:
@kks32 
This is my first pull request. This commit includes the code to import CSS raw data and change it to dataframe.

Commit:
import css data

Commit:
delete cssdata.basedir function

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+# import from my modules
+from dataimportingfunctions import get_data_dir_list
+
+def exp_num_list():
```
You don't need this function, pass this as an argument at runtime

yjchoi1 replied:
What is "pass this as an argument at runtime"? I don't understand what it means. I have made a `.json` file to define variables.

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+    return exp_num_list
+
+
+def relative_density():
```
Again, these are specific to your dataset and must be stored in a file and read, not as a function

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+def to_dataframe(data_dir, exp_num_list=exp_num_list(), drs=relative_density()):
+    """make a list to contain dataframe for the whole data"""
+
+    df_all = []  # make an empty list to contain dataframes
```
dfs or df

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+    return drs
+
+
+def to_dataframe(data_dir, exp_num_list=exp_num_list(), drs=relative_density()):
```
What are you converting to a dataframe?

yjchoi1 replied:
I am converting `.csv` data to a dataframe.

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+    df_all = []  # make an empty list to contain dataframes
+
+    # get the number of trials for each experiment
+    for exp in exp_num_list:
```
A pythonic way to write a for loop is `for experiment in experiments:`

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+
+    # get the number of trials for each experiment
+    for exp in exp_num_list:
+        data_dir_list = get_data_dir_list(exp, basedir=data_dir)  # get dir list of trial files in each exp
```
Why is this a list? Also don't add type to variable names. Remove the `get_` in the function name

yjchoi1 replied:
I made it as a list to get a list of `.cvs` file directories for each experiment.

yjchoi1 replied:
I would like to specify that the purpose of this function is to get a list of data_dirs and returns the list. Is it better to describe it in the function description part rather than doing it in the function name? For me, it is more convenient when I specify in the function name.

yjchoi1 replied:
In addition, the reason that I used the variable type to the variable was because I wanted to distinguish between the following:
- a list containing lists (e.g., [[1,2,3,4,5], [1,2,3,4,1], [3,4,1,2]]). In this case I would like to name it `datas_list`
- a list containing datas (e.g., [1,2,3,4,5]). In this case, I would like to name it `datas`
- a single variable. In this case, I would like to name it `data`

Is there other recommended naming convention?

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+    # get the number of trials for each experiment
+    for exp in exp_num_list:
+        data_dir_list = get_data_dir_list(exp, basedir=data_dir)  # get dir list of trial files in each exp
+        num_exps = len(exp_num_list)  # get num of experiments
```
`num_experiments`

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+        for trial in range(num_trials):
+
+            # make a dataframe for each trial
+            data_dir_trial = data_dir_list[trial]  # get directory of each `trial.csv`
```
You don't need a variable, use it directly in the next line

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+
+            # compute and insert ru at the 4th column of the dataframe
+            # confining pressure of the test is the first (starting) effective vertical stress
+            confining_pressure = df_single.iloc[0, 4]  # get conf pressure of the trial
```
Can't you read it from the header in csv?

yjchoi1 replied:
What is 'it' you are referring to?

yjchoi1 replied:
If it is confining pressure, there is no information about it in the header. I confirmed with Dr. El-Mohtar that the effective vertical stress at the first raw corresponds to the confining pressure of the test.

kks32 replied:
Can't append `ru` to the end of the dataframe, why a specific location like 4?

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+    return df_all
+
+
+def plot_trial(dataframe, expindex, trialindex):
```
This is not plotting, it doesn't have plot.show or plot.save.

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+        axi.set(ylabel=data_col_names[i])
+
+
+def look_into_data(dataframe, timeindex, exp_num_list=exp_num_list()):
```
Rename to `inspect_data`

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+
+def look_into_data(dataframe, timeindex, exp_num_list=exp_num_list()):
+    """
+    Some of data has a irregular time interval and different data points.
```
What does look into even mean?

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,112 @@
+            time_choose = time[timeindex]  # get the time (sec) at the specified index
+            time_interval = time_last/index_last  # get the time interval (sec) between each data point
+
+            print(f"last index and time:{index_last}, {time_last}; interval: {time_interval:.4f}\n{timeindex} time step is: {time_choose} sec")
```
If the function is merely printing info, then say so in the function name `print_experiment_summary`

kks32 commented on dataimportingfunctions.py:
```diff
@@ -0,0 +1,26 @@
+import os
```
Rename file to `import_data.py`

kks32 commented on train.py:
```diff
@@ -0,0 +1,24 @@
+# %%
+import tensorflow as tf
+import numpy as np
+import os
+import json
+from matplotlib import pyplot as plt
+
+from tensorflow.keras.layers import Dense, Dropout, SimpleRNN, LSTM, GRU, Flatten
+from tensorflow.keras import optimizers
+from tensorflow.keras.callbacks import EarlyStopping
+
+import cssdata
```
Only import things you are using

yjchoi1 replied:
These libraries will be used later in the `train.py` file, but I will delete those in this step as you recommended.

kks32 commented on dataimportingfunctions.py:
```diff
@@ -0,0 +1,26 @@
+    For example, when expnum=1, it returns ["MyData/trial-1", "MyData/trial-2", ..., "MyData/trial-n"]
+    """
+    
+    experimentlist = os.listdir(basedir)  # get a list of file names in `basedir`
+    experimentlist = [ex for ex in experimentlist if 'Experiment' in ex]  # only get `Experiment` folders
+    experimentlist = natsorted(experimentlist)  # `natsorted` enables to sort "1, 10, 2, ..." to "1, 2, ..., 10, ..."
```
Merge these three lines together. the variable changes type so many times in this function :(

kks32 commented on dataimportingfunctions.py:
```diff
@@ -0,0 +1,26 @@
+        datadir = os.path.join(expdir, data)  # dir for a data file
+        datadirlist.append(datadir)  # append
+    
+    # sort in a numerical order
```
Why should it be numerically sorted?

yjchoi1 replied:
For better indexing, I thought it would be better to sort in numerical order.

kks32 commented on train.py:
```diff
@@ -0,0 +1,24 @@
+# %% import data and dataframe
+
+cwd = os.getcwd()  # get current working directory
+data_dir = os.path.join(cwd, "rawdata")  # define the path for the rawdata
+
+exp_num_list = cssdata.exp_num_list()  # get exp_num_list that you want to consider (by default, [7, 8, 9, 10])
+drs = cssdata.relative_density()  # get relative density (Dr) data for each trial
```
Should be passed as input arguments at runtime.

kks32 reviewed (changes requested):
Create a proper name to your pull request. A short name preferably

Commit:
update

updated namings, variable declarations, add json file

yjchoi1 reviewed (commented):
@kks32 I have modified the code based on your comments. Thanks a lot for the detailed comments!

kks32 commented on .gitignore:
```diff
@@ -0,0 +1,10 @@
+*.pyc
+*.idea
+*.ipynb
+test.py
```
Why is this file ignored?

kks32 commented on README.md:
```diff
@@ -1 +1,4 @@
\ No newline at end of file
+# lstm-liquefaction
+## Data
+https://drive.google.com/drive/folders/16m0r_nRRXCeiw_k57_EZbpBqJ5X3VywW?usp=sharing \
```
Add a link to the original DS dataset

kks32 commented on input.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "exp_num_list": [7, 8, 9, 10],
```
```suggestion
  "experiment_ids": [7, 8, 9, 10],
```

kks32 commented on input.json:
```diff
@@ -0,0 +1,13 @@
+{
+  "exp_num_list": [7, 8, 9, 10],
+  "dr_exp7": [0.55, 0.54, 0.47, 0.56, 0.52, 0.52, 0.39, 0.52, 0.41, 0.43,
+    0.39, 0.39, 0.33, 0.35, 0.43, 0.35, 0.41, 0.44, 0.53],
```
```suggestion
  "experiments": [
    {
       "exp_id": 7,
       trials: [
         {
           "id": 1,
           "file": "Trial-1_130524_yc1.csv",
           "relative_density": 0.55
         },
         {
           "id": 2,
           "file": "Trial-2_130527_yc1.csv",
           "relative_density": 0.54         
         }
       ]
     }
  ]
```

kks32 commented on train.py:
```diff
@@ -0,0 +1,27 @@
+
+# %% import data and dataframe
+
+cwd = os.getcwd()  # get current working directory
+data_dir = os.path.join(cwd, "rawdata")  # define the path for the rawdata
+inputfile = "input.json"  # name of the `input.json`
+inputfile_dir = os.path.join(cwd, inputfile)  # define the path for the `input.json`
```
This part should be a command-line argument. Use https://docs.python.org/3/library/argparse.html
Get the working directory and the name of the input JSON file. So you don't need all these superfluous code from lines 9 - 12.

kks32 commented on train.py:
```diff
@@ -0,0 +1,27 @@
+with open(inputfile_dir, "r") as f:
+    json_data = json.load(f)
+
+experiments = json_data["exp_num_list"]  # experiment list that you are importing
+dr_exp7 = json_data["dr_exp7"]  # relative density of each trial in experiment 7
+dr_exp8 = json_data["dr_exp8"]  # relative density of each trial in experiment 8
+dr_exp9 = json_data["dr_exp9"]  # relative density of each trial in experiment 9
+dr_exp10 = json_data["dr_exp10"]  # relative density of each trial in experiment 10
+drs = [dr_exp7, dr_exp8, dr_exp9, dr_exp10]  # collect relative densities in a list
```
All these will change once you fix your JSON.

kks32 reviewed (changes requested):
Revise your JSON structure. This is a big PR, until you get comfortable writing proper Python codes. Keep the PRs small.

kks32 commented on train.py:
```diff
@@ -0,0 +1,27 @@
+inputfile_dir = os.path.join(cwd, inputfile)  # define the path for the `input.json`
+
+# open the `input.json` file
+with open(inputfile_dir, "r") as f:
+    json_data = json.load(f)
```
input = json.load(open(input_json_file, "r"))

kks32 commented on train.py:
```diff
@@ -0,0 +1,27 @@
+with open(inputfile_dir, "r") as f:
+    json_data = json.load(f)
+
+experiments = json_data["exp_num_list"]  # experiment list that you are importing
+dr_exp7 = json_data["dr_exp7"]  # relative density of each trial in experiment 7
+dr_exp8 = json_data["dr_exp8"]  # relative density of each trial in experiment 8
+dr_exp9 = json_data["dr_exp9"]  # relative density of each trial in experiment 9
+dr_exp10 = json_data["dr_exp10"]  # relative density of each trial in experiment 10
+drs = [dr_exp7, dr_exp8, dr_exp9, dr_exp10]  # collect relative densities in a list
```
```suggestion

```

kks32 replied:
All these should go into `experiments`

kks32 commented on train.py:
```diff
@@ -0,0 +1,27 @@
+drs = [dr_exp7, dr_exp8, dr_exp9, dr_exp10]  # collect relative densities in a list
+
+# get dataframes for all trials
+dfs = cssdata.csv_to_dataframe(basedir=data_dir, experiments=experiments, drs=drs)
```
Pass only two arguments `working_dir` and `experiments`. 
`experiments` should have a file location, experiment id and relative density.

kks32 commented on import_data.py:
```diff
@@ -0,0 +1,24 @@
+import os
+from natsort import natsorted  # Sort 1, 10, 2, 3 to 1, 2, 3, 10
+
+def data_dirs(expnum, basedir):
+    """
+    get `.csv` data directory list for specified experiments
+    For example, when expNum=1, it returns ["MyData/trial-1", "MyData/trial-2", ..., "MyData/trial-n"]
+    """
+
+    experimentfiles = natsorted([file for file in os.listdir(basedir) if 'Experiment' in file])  # only get `Experiment` folders
+    experimentfile = experimentfiles[expnum - 1]  # get the file name that you are targeting
+    experimentdir = os.path.join(basedir, experimentfile)  # get the data file dir with `join`
+    datanames = os.listdir(experimentdir)  # get the data file list
+
+    # get the data file directory lists for a specified `expNum`
+    datadirs = []  # make an empty list to contain directories
+    for dataname in datanames:
+        datadir = os.path.join(experimentdir, dataname)  # dir for a data file
+        datadirs.append(datadir)  # append
+
+    # sort in a numerical order
+    datadirs = natsorted(datadirs)
```
You need to relate each of your trial file names to the corresponding relative densities.

kks32 replied:
This whole function is unnecessary.

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,88 @@
+        for trial in range(num_trials):
+
+            # make a dataframes for each trial
+            df_single = pd.read_csv(datadirs[trial], header=5)  # make `.csv` to pandas `df`
```
Don't use magic numbers, how do you know it's 5?

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,88 @@
+            df_single = pd.read_csv(datadirs[trial], header=5)  # make `.csv` to pandas `df`
+
+            # insert Dr values at the first column of the dataframes
+            df_single.insert(0, "Dr [%]", drs[experiment - 7][trial])
```
Again what's with the magic number 7?

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,88 @@
+    return dfs
+
+
+def plot_trial(dataframes, expindex, trialindex):
```
If this function is supposed to plot the result of a single trial, shouldn't you just pass the trail object? Don't do figuring out where the trial data is inside this function. This function should get the trial data and plot - that's all. No "exp_index" or "trial_index"

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,88 @@
+
+        for i in range(len_data_list):
+            print(f"*Trial-{i+1}")
+            df = dataframes[k][i]  # load dataframes for the specified exp-trial
```
The way you have organized your dataframe is not correct. Once you change your JSON, we should restructure your DF.

kks32 reviewed (changes requested):
Revise your JSON structure.

Commit:
change README.md

yjchoi1 committed:
json file update

yjchoi1 committed:
update functions based on json

1)updated cssdata.csv_to_dataframe to be based on json and confirmed the result. 2)removed all the magic numbers in cssdata.csv_to_datafram

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,51 @@
+    for experiment in experiments:
+        # iterate over experiments and find the experiment that corresponds to `exp_id` specified.
+        if experiment['exp_id'] == exp_id:
+            trials = experiment['trials']
```
trials doesn't exist outside of this if condition. You need to move the rest of the for trial in trials under this if condition.

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,51 @@
+    :return: dataframe
+    """
+    experiments = input['experiments']
+    for experiment in experiments:
```
```suggestion
    for experiment in input['experiments']:

```

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,51 @@
+    :param num_headers: number of headers in csv file
+    :return: dataframe
+    """
+    experiments = input['experiments']
```
```suggestion

```

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,51 @@
+    experiments = input['experiments']
+    for experiment in experiments:
+        # iterate over experiments and find the experiment that corresponds to `exp_id` specified.
+        if experiment['exp_id'] == exp_id:
```
Define an empty list of trials, otherwise trials is undefined.

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,51 @@
+                file = trial['file']
+                relative_density = trial['relative_density']
+                # convert csv to dataframe
+                df = pd.read_csv(f"./rawdata/Experiment-{exp_id}/{file}", header=num_headers)
```
Why are you still adding `./rawdata/Experiment-stuff!` shouldn't this be read from the input JSON?

kks32 commented on import_data.py:
```diff
@@ -0,0 +1,24 @@
+import os
+from natsort import natsorted  # Sort 1, 10, 2, 3 to 1, 2, 3, 10
+
+def data_dirs(expnum, basedir):
```
Is this function being used anywhere?

kks32 commented on train.py:
```diff
@@ -1,27 +1,13 @@
 
 # get dataframes for all trials
-dfs = cssdata.csv_to_dataframe(basedir=data_dir, experiments=experiments, drs=drs)
+cssdata.csv_to_dataframe(input=input, exp_id=7, trial_id=1)
```
What about the return value?

yjchoi1 committed:
update cssdata

update csv_to_dataframe, plot_trial

kks32 commented on cssdata.py:
```diff
@@ -0,0 +1,50 @@
+    return df
+
+
+def plot_trial(input, exp_id, trial_id, ncols=1, figsize=(13, 15)):
```
Why don't you pass the dataframe instead of parsing it inside the function

kks32 commented on train.py:
```diff
@@ -0,0 +1,15 @@
+input = json.load(open("input.json", "r"))
+
+# get dataframes for all trials
+cssdata.csv_to_dataframe(input=input, exp_id=8, trial_id=3)
```
What are you doing with the returned dataframe?

yjchoi1 committed:
update json

kks32 commented on cssdata.py:
```diff
@@ -23,7 +23,7 @@ def csv_to_dataframe(input, exp_id, trial_id, num_headers=5):
                     relative_density = trial['relative_density']
                     # convert csv to dataframe
-                    df = pd.read_csv(f"./rawdata/Experiment-{exp_id}/{file}", header=num_headers)
+                    df = pd.read_csv(file, header=num_headers)
```
Nice!

kks32 commented on input.json:
```diff
@@ -166,237 +166,237 @@
         {
           "id": 18,
-          "file": "Trial-18_130222.csv",
+          "file": "./rawdata/Experiment-9/Trial-18_130222.csv",
```
I don't think you need `./` in front of the rawdata. You can just do `rawdata/Experiment...`

yjchoi1 committed:
drop ./ in jason

yjchoi1 committed:
delete importdata.py

kks32 commented on train.py:
```diff
@@ -0,0 +1,15 @@
+input = json.load(open("input.json", "r"))
+
+# get dataframes for a trial
+cssdata.csv_to_dataframe(input=input, exp_id=7, trial_id=3)
```
capture dataframe that is returned

kks32 reviewed (changes requested):
Just a couple more comments left to be resolved.

yjchoi1 committed:
update preparedata.py

kks32 commented:
Nice work @yjchoi1 !!!
